mars/oscar/backends/ray/communication.py

```diff
@@ -0,0 +1,340 @@
+        try:
+            # Wait on ray object ref
+            object_ref = await self._in_queue.get()
+            return await object_ref
+        except RuntimeError as e:
+            if not self._closed.is_set():
+                raise e
+
+
+class RayOneWayActorChannel(RayChannelBase):
+    """A channel from ray actor to ray driver. Since ray actor can't call ray actor,
+     we use ray call reply for channel recv. Note that there can't be multiple
+     channel message send for one received message, or else it will be taken as next
+     message's reply.
+    """
+    __slots__ = '_out_queue', '_msg_recv_counter', '_msg_sent_counter'
+
+    def __init__(self,
+                 local_address: str = None,
+                 channel_index: int = None,
+                 channel_id: ChannelID = None,
+                 compression=None):
+        super().__init__(local_address, None, channel_index, channel_id, compression)
+        self._out_queue = asyncio.Queue()
+        self._msg_recv_counter = 0
+        self._msg_sent_counter = 0
+
+    @implements(Channel.send)
+    async def send(self, message: Any):
+        if self._closed.is_set():  # pragma: no cover
+            raise ChannelClosed('Channel already closed, cannot send message')
+        # Current process is ray actor, peer is ray driver.
+        # We can't ray call to ray driver for message send, so we use ray call reply
+        # to send message to ray driver.
+        # Not that we can send once for every read message in channel, other else
+        # it will be taken as other message's reply.
+        await self._out_queue.put(message)
+        self._msg_sent_counter += 1
+        assert self._msg_sent_counter <= self._msg_recv_counter,\
+            "One way channel doesn't support multiple reply for one message."
+
+    @implements(Channel.recv)
+    async def recv(self):
+        if self._closed.is_set():  # pragma: no cover
+            raise ChannelClosed('Channel already closed, cannot write message')
+        try:
+            return await self._in_queue.get()
+        except RuntimeError:
+            if self._closed.is_set():
+                pass
+
+    async def __on_ray_recv__(self, message):
+        """This method will be invoked when current process is a ray actor rather than a ray driver"""
+        self._msg_recv_counter += 1
+        await self._in_queue.put(message)
+        return await self._out_queue.get()
+
+
+@register_server
+class RayServer(Server):
+    __slots__ = '_closed', '_address', '_channels', '_tasks'
+
+    # Multiple instances for ray local mode
+    _address_to_instances: Dict[str, "RayServer"] = dict()
+    scheme = 'ray'
+
+    def __init__(self, address, channel_handler: Callable[[Channel], Coroutine] = None):
+        super().__init__(DEFAULT_DUMMY_RAY_ADDRESS, channel_handler)
+        self._address = address
+        self._closed = asyncio.Event()
+        self._channels: Dict[ChannelID, RayTwoWayChannel] = dict()
```

Comment:
The server holds two different channels looks a bit weird, split into `DriverServer` and `ActorServer` seems better, how do you think? @qinxuye

Reply:
Because ray driver cab't be called by ray actors, ray driver doesn't have a`Server`

Reply:
When connection happens, the channel type could be decided by the client, for example, actor or driver, and then create the corresponding channel.

Reply:
The point is that the ray actor will receive the requests from both the ray driver and ray actors. It needs different channels to send send messages back . When the request is from the ray driver, it use `RayOneWayActorChannel`;  When the request is from the ray actor, it use `RayTwoWayChannel`. The channel is decided by peer process address. If peer process is ray driver ,the address will be None, or else it will be the actor's name.